rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,826 @@
+during the OpenXLA Dev Summit in April 2023.
+
+**Proposal:** This RFC proposes to ratify this convention as a StableHLO design
+principle, so that the folklore consensus becomes codified in the StableHLO
+specification.
+
+**Discussion:** A) The rationale for this proposal is that lower-level
+abstractions (e.g. HLO or Linalg) typically don't want to concern themselves
+with shape checks - they are focused on generating high-performance code, so
+they want shape checks to be handled at a higher level, i.e. at StableHLO
+or above.
+
+Furthermore, different producers have different requirements and different
+preferences for expressing shape checks (e.g. JAX's type system enables it to
+need fewer checks than TensorFlow's type system), so a specific way of
+performing shape checks don't look like something that should be standardized
+at the StableHLO level either.
+
+B\) From the implementation standpoint, this proposal would involve updating
+[the "Errors" section](../docs/spec.md#errors) of the specification, and
+changing the StableHLO dialect to implement the `ConditionallySpeculatable`
+interface, prohibiting speculation for StableHLO ops that involve dynamism
+([documentation](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/)).
+
+<!-- markdownlint-disable line-length -->
+## <a id="p3" /> (P3) Ratify the existing convention for relaxed constraints already implemented in the StableHLO dialect
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** At the moment, the StableHLO dialect supports
+[relaxed constraints](https://github.com/openxla/stablehlo/blob/8993ad54839add6648b88801f1d223b7f9bc2e58/stablehlo/dialect/Base.cpp#L102-L120)
+that were inherited from the MHLO dialect. For example, the code below is valid
+in the StableHLO dialect:
+
+```mlir
+func.func @main(%arg0: tensor<?xf32>, %arg1: tensor<1xf32>) {
+  %0 = stablehlo.add %arg0, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<?xf32>
+  %1 = stablehlo.add %arg0, %arg0 : (tensor<?xf32>, tensor<?xf32>) -> tensor<1xf32>
+  %2 = stablehlo.add %arg0, %arg1 : (tensor<?xf32>, tensor<1xf32>) -> tensor<?xf32>
+  %3 = stablehlo.add %arg0, %arg1 : (tensor<?xf32>, tensor<1xf32>) -> tensor<1xf32>
+  %4 = stablehlo.add %arg1, %arg0 : (tensor<1xf32>, tensor<?xf32>) -> tensor<?xf32>
+  %5 = stablehlo.add %arg1, %arg0 : (tensor<1xf32>, tensor<?xf32>) -> tensor<1xf32>
+  %6 = stablehlo.add %arg1, %arg1 : (tensor<1xf32>, tensor<1xf32>) -> tensor<?xf32>
+  %7 = stablehlo.add %arg1, %arg1 : (tensor<1xf32>, tensor<1xf32>) -> tensor<1xf32>
+  func.return
+}
+```
+
+More formally, these relaxed constraints generalize the constraints that are
+already documented in the StableHLO specification. If a constraint for a
+specific operation cannot be evaluated at compile time because it involves an
+unknown rank or an unknown dimension size, it gets deferred until the run time
```

Comment:
Would be good to clarify what "cannot be evaluated" means.

If there exists some value of `?` where a constraint could succeed, I understand deferring to runtime. I'm unsure about how we should handle cases where no value of `?` would permit verification to succeed, or if this is even feasible.

Reply:
For an simple example of the above comment, consider the following. StableHLO ConcatenateOp requires that all non-concatenated dimensions be equivalent, making `[2, 1] . [2, 5]` with concat `dim = 1` valid since non-concatenated dims are both `2`.

Enter dynamism:

```
%0 = stablehlo.concatenate %arg0, %arg1, dim = 1 : (tensor<?x1xf32>, tensor<2x1xf32>) -> tensor<?x2xf32>
%1 = stablehlo.concatenate %arg0, %arg1, %arg2, dim = 1 : (tensor<?x1xf32>, tensor<2x1xf32>, tensor<3x1xf32>) -> tensor<?x3xf32>
```

- For `%0`, this is valid if `? == 2`
- For `%1`, no value of `?` satisfies `? == 2 == 3`

Thinking more about my earlier comment, while erroring on `? == 2 == 3` is certainly a more precise representation, I'm not sure _how useful_ this error is in practice..

I see two use cases for dynamically shaped programs:
1. Backend can evaluate dynamic shapes (mostly interpreters).
2. Backend refines shapes to run static program (compiler).

Users of (1), will need to implement identical verification regardless (in this case concat dimension size verification) to ensure during runtime that the provided value is valid.

Users in (2) will get static verification after refinement.

The use case I can imagine being interested in stricter verification would be serializing/shipping dynamic unrefined models, as (some of) these models wouldn't error until they attempt to be executed by a consumer instead of as they are serialized by the producer. Note that even with strict verification it is still possible to refine a program using invalid values and still fail verification, the only difference in strict verification is that "some value of `?` exists which can satisfy a constraint".

Reply:
"Would be good to clarify what "cannot be evaluated" means". This RFC is using the meaning of evaluation that was informally established in https://github.com/openxla/stablehlo/blob/main/docs/spec.md#formulas.

In a nutshell, evaluation of a constraint takes a constraint written in the StableHLO specification formula language, resolves names in the constraint using the rules from the "Names" section, applies some peculiar semantics described in the "Formulas" section (e.g. ellipses) and produces one of the supported values described in the "Values" section.

I mentioned above that this notion of evaluation hasn't been formalized. When writing up the relevant parts of the StableHLO spec, I was thinking that we can assume that the readers would be familiar with mainstream programming languages like Python, which would allow them to extend their intuition about evaluation in these languages to formulas used in the StableHLO specification.

To make things very precise, we could formalize the notion of evaluation - both for StableHLO formulas and StableHLO programs - using one of the frameworks popular in theoretical works on programming languages. E.g. defining small-step operational semantics could be one way to approach this. On a related note, we actually have a ticket for formalizing evaluation of StableHLO programs (#484), but I haven't yet gottten around to doing that.

"I'm unsure about how we should handle cases where no value of ? would permit verification to succeed, or if this is even feasible". I think this is an interesting idea. I like that it would allow us to reject more invalid StableHLO programs at compile time. 

However, what's unclear to me is how to frame the formula evaluation rules to accommodate this proposal. That would probably mean that Python/C++/etc-style evaluation semantics is out of the question, and we would need some form of solvers?